AirCasting/Utils/Connection/BluetoothConnection.swift

```diff
@@ -0,0 +1,53 @@
+import SwiftUI
+import CoreData
+import Resolver
+
+final class BluetoothConnectionProtector: Connectable, SessionsFetchable {
+    
+    enum BluetoothConnectionProtectorError: Error {
+        case alreadyConnected
+        case readError(Error)
+    }
+    
+    @Injected private var persistenceController: PersistenceController
+    
+    func fetchSessions(constrained: Database.Constraint, completion: @escaping (Result<[Database.Session], Error>) -> Void) {
+        let context = persistenceController.editContext
+        context.performAndWait {
+            let request: NSFetchRequest<SessionEntity> = SessionEntity.fetchRequest()
+            if case .predicate(let predicate) = constrained {
+                request.predicate = predicate
+            }
+            do {
+                let fetched = try context.fetch(request)
+                let databaseObjects = fetched.map(Database.Session.init(coreDataEntity:))
+                completion(.success(databaseObjects))
+            } catch {
+                completion(.failure(BluetoothConnectionProtectorError.readError(error)))
+            }
+        }
+    }
+    
+    func isAirBeamAvailableForNewConnection() -> Result<Void, Error> {
+        var result: Result<Void, Error>!
+        let predicate = NSPredicate(format: "deviceType == %@ AND type == %@ AND (status == %li || status == %li || status == %li)",
+                                    "AIRBEAM3",
+                                    SessionType.mobile.rawValue,
+                                    SessionStatus.RECORDING.rawValue,
+                                    SessionStatus.DISCONNECTED.rawValue,
+                                    SessionStatus.NEW.rawValue)
+        fetchSessions(constrained: .predicate(predicate)) { predicated in
+            switch predicated {
+            case .success(let sessions):
+                if sessions.count > 0 {
+                    result = .success(())
+                } else {
+                    result = .failure(BluetoothConnectionProtectorError.alreadyConnected)
+                }
```

Comment:
why not `result = sessions.count > 0 ? .success(()) : .faiulre(...)`?

Also - isn't that flipped? It will return "yes, we can connect" if it **finds** ongoing sessions, won't it?

Reply:
I will, but please if you could comment it as much as possible - it is still a bit tough for me